### ISSUE TITLE
RFE: Add support for importing CA certificate (needed to register against Katello/Satellite6)

### ISSUE THREAD

**Issue by epienbroek** [projectatomic/adb-vagrant-registration#38]
I'm trying to use this plugin to allow vagrant boxes to be registered against our internal Katello server. To achieve this I added the following parameters to the Vagrantfile:
```
    config.registration.serverurl = 'https://katello.foo.bar/rhsm'
    config.registration.org = 'Default Organization'
    config.registration.activationkey = 'ak-rhel-guest'
```

This currently fails with the following error:
```
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

cmd=$(subscription-manager register --serverurl='https://katello.foo.bar/rhsm' --org='Default Organization' --activationkey='ak-rhel-guest' --force); if [ "$?" != "0" ]; then echo $cmd | grep 'This system is already registered' || (echo $cmd 1>&2 && exit 1) ; fi

Stdout from the command:



Stderr from the command:

Error: CA certificate for subscription service has not been installed.
```

Could you please add support for supplying a CA certificate or the location of the CA certificate RPM (for example http://katello.foo.bar/pub/katello-ca-consumer-latest.noarch.rpm) and make sure it is configured inside the guest properly before trying to do the actual 'subscription-manager register' ?

**Comment by strzibny** [comment]
Would the following work?
```
    config.registration.serverurl = 'https://katello.foo.bar/rhsm'
    config.registration.org = 'Default Organization'
    config.registration.activationkey = 'ak-rhel-guest'
    config.registration.ca_cert = '/path/to/crt'
```
Certificate would be then copying to the guest before attempting to register.

Is there somewhere Katello's ca consumer RPM? So I can check what it contains?

**Comment by epienbroek** [comment]
The Katello CA RPM contains just one file, /etc/rhsm/ca/katello-server-ca.pem. This CA file alone is also published by default on the katello server at http://katello.foo.bar/pub/katello-server-ca.crt.

It doesn't matter to me whether the vagrant-registration configuration points to the location of the RPM (local/remote) or the location of the CA file (local/remote). Either way is sufficient for me.

**Comment by strzibny** [comment]
Just to confirm, is the `/etc/rhsm/ca/katello-server-ca.pem` needed? If yes, we would have to let people specify the CA cert destination.

**Comment by epienbroek** [comment]
The default subscription-manager configuration contains the parameter ```ca_cert_dir = /etc/rhsm/ca/``` so I would expect that it automatically processes all files in this folder.

To confirm this I just tried to put the Katello CA certificate in a file named /etc/rhsm/foo.pem (instead of installing the Katello CA RPM) and execute subscription-manager using strace:
```
# strace -f subscription-manager register --org="Default Organization" --name="node1.foo.bar" --activationkey="ak-rhel-guest" 2>&1 | grep /etc/rhsm
open("/etc/rhsm/rhsm.conf", O_RDONLY)   = 7
open("/etc/rhsm/ca/", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 4
open("/etc/rhsm/ca/foo.pem", O_RDONLY)  = 4
open("/etc/rhsm/ca/redhat-uep.pem", O_RDONLY) = 4
open("/etc/rhsm/ca/candlepin-stage.pem", O_RDONLY) = 4
open("/etc/rhsm/ca/", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 4
open("/etc/rhsm/ca/foo.pem", O_RDONLY)  = 4
open("/etc/rhsm/ca/redhat-uep.pem", O_RDONLY) = 4
open("/etc/rhsm/ca/candlepin-stage.pem", O_RDONLY) = 4
```
As can be seen in this output, subscription-manager automatically processes all files in the folder /etc/rhsm/ca so the name of the CA cert destination doesn't matter.

**Comment by strzibny** [comment]
I would like you to test my changes in https://github.com/projectatomic/adb-vagrant-registration/pull/42

You can build and install this plugin with:
```
git checkout uploading-ca-cert
gem build vagrant-registration.gemspec
vagrant plugin install vagrant-registration-0.0.19.gem
```

Then I will be able to merge it. Thanks a lot.

**Comment by epienbroek** [comment]
Thanks for implementing this request! I just tried the PR and can confirm that it works.
There's only one minor gotcha: the filename of the CA certificate must end with the extension '.pem' otherwise subscription-manager silently ignores the file. But other than that it works great!

Here the vagrant output:
```
Bringing machine 'lvs1' up with 'libvirt' provider...
==> lvs1: Creating image (snapshot of base box volume).
==> lvs1: Creating domain with the following settings...
==> lvs1:  -- Name:              puppet_conf_lvs1
==> lvs1:  -- Domain type:       kvm
==> lvs1:  -- Cpus:              1
==> lvs1:  -- Memory:            512M
==> lvs1:  -- Base box:          rhel-5.8-x86_64
==> lvs1:  -- Storage pool:      default
==> lvs1:  -- Image:             /var/lib/libvirt/images/puppet_conf_lvs1.img
==> lvs1:  -- Volume Cache:      default
==> lvs1:  -- Kernel:            
==> lvs1:  -- Initrd:            
==> lvs1:  -- Graphics Type:     vnc
==> lvs1:  -- Graphics Port:     5900
==> lvs1:  -- Graphics IP:       127.0.0.1
==> lvs1:  -- Graphics Password: Not defined
==> lvs1:  -- Video Type:        cirrus
==> lvs1:  -- Video VRAM:        9216
==> lvs1:  -- Keymap:            en-us
==> lvs1:  -- Command line : 
==> lvs1: Creating shared folders metadata...
==> lvs1: Starting domain.
==> lvs1: Waiting for domain to get an IP address...
==> lvs1: Waiting for SSH to become available...
    lvs1: 
    lvs1: Vagrant insecure key detected. Vagrant will automatically replace
    lvs1: this with a newly generated keypair for better security.
    lvs1: 
    lvs1: Inserting generated public key within guest...
    lvs1: Removing insecure key from the guest if it's present...
    lvs1: Key inserted! Disconnecting and reconnecting using new SSH key...
==> lvs1: Setting hostname...
==> lvs1: Configuring and enabling network interfaces...
==> lvs1: Registering box with vagrant-registration...
==> lvs1: Uploading CA certificate from katello-server-ca.pem...
==> lvs1: Rsyncing folder: /home/erik/puppet_conf/ => /vagrant
...
```

and inside the guest:
```
[root@lvs1 ~]# subscription-manager status
+-------------------------------------------+
   System Status Details
+-------------------------------------------+
Overall Status: Current
```

**Comment by strzibny** [comment]
Yes, I know. But I decided to keep the original name that people will use rather then renaming it. Perhaps there could be one more warning that the file would be ignored. Good idea. Thanks for testing.

**Comment by strzibny** [comment]
In the end I decided that if the file does not end with '.pem' extension, this extension is automatically added. It's now merged in master and I will release 1.0 soon.

**Comment by epienbroek** [comment]
During testing I've found out that an additional change is needed. When trying to use yum after the registration is done yum currently assumes that the CA certificate for the automatically configured yum repositories is /etc/rhsm/ca/redhat-uep.pem.This is not the case when registering against a Katello/Satellite 6 environment and results in errors like:
```
# yum install iptables
Loaded plugins: product-id, subscription-manager
Traceback (most recent call last):
  File "/usr/bin/yum", line 29, in ?
    yummain.user_main(sys.argv[1:], exit_code=True)
  File "/usr/share/yum-cli/yummain.py", line 309, in user_main
    errcode = main(args)
  File "/usr/share/yum-cli/yummain.py", line 178, in main
    result, resultmsgs = base.doCommands()
  File "/usr/share/yum-cli/cli.py", line 345, in doCommands
    self._getTs(needTsRemove)
<<snip>>
  File "/usr/lib64/python2.4/site-packages/M2Crypto/SSL/Connection.py", line 167, in connect_ssl
    return m2.ssl_connect(self.ssl, self._timeout)
M2Crypto.SSL.SSLError: certificate verify failed
```

The parameter repo_ca_cert in /etc/rhsm/rhsm.conf also needs to be updated so that it points to the correct CA certificate (instead of the default '%(ca_cert_dir)sredhat-uep.pem'). This can be done after the 'subscription-manager register' has been performed already

**Comment by strzibny** [comment]
So you believe the best way is to always make sure that `repo_ca_cert` is set to our expected path in `/etc/rhsm/rhsm.conf`?

**Comment by epienbroek** [comment]
Correct
This is also what the Katello CA RPM does using the following RPM %post script:
```
[root@katello ~]# rpm --query --scripts katello-ca-consumer-katello.foo.bar-1.0-1.noarch
postinstall scriptlet (using /bin/sh):
#!/bin/bash
#
# Copyright 2013 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public License,
# version 2 (GPLv2). There is NO WARRANTY for this software, express or
# implied, including the implied warranties of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
# along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
#
# Red Hat trademarks are not licensed under GPLv2. No permission is
# granted to use or replicate Red Hat trademarks that are incorporated
# in this software or its documentation.
#
# Configures rhsm on client. Called from the certificate RPM.
#

KATELLO_SERVER=katello.foo.bar
PORT=443
BASEURL=https://$KATELLO_SERVER/pulp/repos

PREFIX=/rhsm
CFG=/etc/rhsm/rhsm.conf
CFG_BACKUP=$CFG.kat-backup

# Get version of RHSM
RHSM_V="`rpm -q --queryformat='%{VERSION}' subscription-manager 2> /dev/null | tr . ' '`"
if test $? != 0 ; then
  RHSM_V="0 0 0"
fi
declare -a RHSM_VERSION=($RHSM_V)

# exit on non-RHEL systems or when rhsm.conf is not found
test -f $CFG || exit
type -P subscription-manager >/dev/null || type -P subscription-manager-cli >/dev/null || exit

# backup configuration during the first run
test -f $CFG_BACKUP || cp $CFG $CFG_BACKUP

# configure rhsm
# the config command was introduced in rhsm 0.96.6
# fallback left for older versions
if test ${RHSM_VERSION[0]:-0} -gt 0 -o ${RHSM_VERSION[1]:-0} -gt 96 -o \( ${RHSM_VERSION[1]:-0} -eq 96 -a ${RHSM_VERSION[2]:-0} -gt 6 \); then
  subscription-manager config \
    --server.hostname="$KATELLO_SERVER" \
    --server.prefix="$PREFIX" \
    --server.port="$PORT" \
    --rhsm.repo_ca_cert="%(ca_cert_dir)skatello-server-ca.pem" \
    --rhsm.baseurl="$BASEURL"
else
  sed -i "s/^hostname\s*=.*/hostname = $KATELLO_SERVER/g" $CFG
  sed -i "s/^port\s*=.*/port = $PORT/g" $CFG
  sed -i "s|^prefix\s*=.*|prefix = $PREFIX|g" $CFG
  sed -i 's|^repo_ca_cert\s*=.*|repo_ca_cert = %(ca_cert_dir)skatello-server-ca.pem|g' $CFG
  sed -i "s|^baseurl\s*=.*|baseurl=$BASEURL|g" $CFG
fi

if grep --quiet full_refresh_on_yum $CFG; then
  sed -i "s/full_refresh_on_yum\s*=.*$/full_refresh_on_yum = 1/g" $CFG
else
  full_refresh_config="#config for on-premise management\nfull_refresh_on_yum = 1"
  sed -i "s/baseurl.*/&\n\n$full_refresh_config/g" $CFG
fi

# restart goferd if it is installed and running
[ -f /etc/init.d/goferd ] && \
  service goferd status >/dev/null && \
  service goferd restart >/dev/null 2&>1

exit 0

# vim:sw=2:ts=2:et:
```

**Comment by strzibny** [comment]
@epienbroek Can you try the following branch? https://github.com/projectatomic/adb-vagrant-registration/tree/repo_ca_cert-option

**Comment by strzibny** [comment]
@epienbroek do you have some time to test if the change I made is enough? I tested that it properly changes the option but I would like to confirm that it works for you before I publish a revision. Thanks.

**Comment by epienbroek** [comment]
Thanks for the fix. I just tested it and indeed resolves the issue:
```
==> lvs1: Setting repo_ca_cert option in /etc/rhsm/rhsm.conf...
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: sed -i 's|^repo_ca_cert *=.*|repo_ca_cert = /etc/rhsm/ca/katello-server-ca-infoplus-ot.pem|g' /etc/rhsm/rhsm.conf (sudo=true)
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: echo; printf $SSH_AUTH_SOCK (sudo=false)
DEBUG ssh: stdout: 

DEBUG ssh: stdout: /tmp/ssh-QWkFKz2567/agent.2567
DEBUG ssh: stdout: 
DEBUG ssh: Exit status: 0
```

```
$ vagrant ssh lvs1
Last login: Thu Nov 26 10:56:12 2015 from 192.168.121.1
[vagrant@lvs1 ~]$ sudo su -
[root@lvs1 ~]# yum repolist
Loaded plugins: product-id, subscription-manager
repo id                                                                                    repo name                                                                                                         status
rhel-5-server-rpms                                                                         Red Hat Enterprise Linux 5 Server (RPMs)                                                                          16,925
repolist: 16,925
```

**Comment by strzibny** [comment]
This is now released as 1.0.1: https://github.com/projectatomic/adb-vagrant-registration/releases/tag/v1.0.1
